### ISSUE TITLE
Question about loss function

### ISSUE THREAD

**Issue by whynot3333** [CRIPAC-DIG/AUG-MAE#4]
Thank you for your fascinating work! I’m new to MGAE and have a question about the loss function. Regarding Equation 12, the corresponding code is below:
`        loss = self.criterion(x_rec_mask, x_init_mask) + u_loss
        num_all_noeds = mask_prob[:,1].sum() + mask_prob[:,0].sum()
        loss_mask = -self.mask_criterion(x_rec_mask, x_init_mask) + args.belta*(torch.tensor([1.]).to(g.device)/torch.sin(torch.pi/num_all_noeds*(mask_prob[:,0].sum())))`
 Under the adversarial training framework, is the generator’s loss function the negative of the main model’s loss function? Why is this approach valid? Additionally, concerning the regularization loss function, Equation 12 uses a subtraction sign, why does the code use an addition sign instead?

**Comment by AzureLeon1** [comment]
> Thank you for your fascinating work! I’m new to MGAE and have a question about the loss function. Regarding Equation 12, the corresponding code is below: ` loss = self.criterion(x_rec_mask, x_init_mask) + u_loss num_all_noeds = mask_prob[:,1].sum() + mask_prob[:,0].sum() loss_mask = -self.mask_criterion(x_rec_mask, x_init_mask) + args.belta*(torch.tensor([1.]).to(g.device)/torch.sin(torch.pi/num_all_noeds*(mask_prob[:,0].sum())))` Under the adversarial training framework, is the generator’s loss function the negative of the main model’s loss function? Why is this approach valid? Additionally, concerning the regularization loss function, Equation 12 uses a subtraction sign, why does the code use an addition sign instead?

Thanks for your interest in our work and for the questions!

**1. About the adversarial objective**
- The **main model** tries to **minimize** the loss to improve reconstruction quality.
- The **mask generator**, on the contrary, tries to **maximize** the same objective, forcing the main model to learn more **robust and informative representations**.
- This process creates a **co-evolution** between the two components, similar to the dynamics observed in **adversarial learning literature**.
*This adversarial min–max optimization paradigm follows the same philosophy as the original GAN formulation introduced in [Goodfellow et al., “Generative Adversarial Nets” (NeurIPS 2014)](https://arxiv.org/pdf/1406.2661)*.

**2. About the sign in Equation 12 vs. implementation**
In **Equation 12**, the generator’s objective is formulated as a **maximization** problem. However, in **PyTorch**, the optimizer (e.g., Adam) always performs **minimization** by default.
To make the implementation consistent with the theoretical objective:
- We **negate the original objective** in the code.
- Minimizing $-\mathcal{L}$ is mathematically equivalent to maximizing $\mathcal{L}$, so the sign in the code appears opposite to the paper’s equation.
- This technique is commonly used in adversarial training and GAN implementations to align with gradient descent optimizers.
So, the code minimizes the **negative form of Equation 12**, which is equivalent to **maximizing the original objective** defined in the paper.
